### DISCUSSION TITLE
End of support of Mac OS 10.15

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/grandorgue#1149] [General]
We inform that since 30 aug 2022 we won't be able to build GrandOrgue for OSx 10.15 according to [the Github annonce](https://github.blog/changelog/2022-07-20-github-actions-the-macos-10-15-actions-runner-image-is-being-deprecated-and-will-be-removed-by-8-30-22/). After this date the oldest supported OSx version will be 11.

You are still able to build GrandOrgue from sources on your legacy Mac youself, but it won't be feasible to perform automatic builds on Github.

Any ideas of continuing automatic builds for old OSx versions are welcome.

**Comment by willeke1234** [comment]
Just an idea: specify the deployment target [MACOSX_DEPLOYMENT_TARGET](https://cmake.org/cmake/help/latest/envvar/MACOSX_DEPLOYMENT_TARGET.html).

I can't test because of #1135.

Edit: found a workaround for #1135 but now GrandOrgue crashes because wxWidgets 3.2.0 doesn't work on macOS 10.15.

Correction: Homebrew doesn't support `MACOSX_DEPLOYMENT_TARGET` and wxWidgets 3.2.0, built for macOS 11, doesn't work on macOS 10.15.
https://github.com/Homebrew/discussions/discussions/370

**Comment by willeke1234** [comment]
I think most people aren't able to build GrandOrgue from sources. Better use an older version.
What is the last release of GrandOrgue running on macOS Catalina 10.15?

**Reply by oleg68** [reply]
> What is the last release of GrandOrgue running on macOS Catalina 10.15?

If we don't find a way for buildung for 10.15, I think, 3.7.0 will be the last one.

**Reply by tux-linux** [reply]
Personnally, I built GO from source for my Mac Pro 2008 running El Capitan (10.11). It was kind of a painful process, especially given how old this OS is, but it worked and I'm able to use it now, whilst the precompiled version provided at the time would cause a segfault/app crash and I could not run it.
So there are some people out there that do build it :)

**Comment by oleg68** [comment]
Unfortunally, I'm not an expert with OSx and how to build an application for an old OSx version using a newer one. If it was possible, I'd add this to the standard github build of GrandOrgue

**Comment by midimusic** [comment]
My opinion:
When Apple changed from Mohave (10.14) to Catalina (10.15) they removed all possibilities of using 32 bit applications and only 64 bit applications are possible. They also rearranged the filing system structure which added further problems. Now they have further changed so that Intel processors are not directly supported.

All of this has convinced me that I will never buy another apple product and have gone back to Mojave in irder to gain the best compatability with older applications (including GO). The binary version of GO for Mojave on Sourceforge works perfectly well and using this overcomes all the problems of inability to compile from source.

It is more than likely that the latest version of GO will compile on Mojave but "brew" will not work in sudo mode and cannot properly access some folders it needs to write to. If the permissions of these folders are changed then it does work.

csw900

**Comment by oleg68** [comment]
I've just compiled GrandOrgue on OSx 11.2 with `-DCMAKE_OSX_DEPLOYMENT_TARGET=10.10'. The application runs on OSx 10.15, but then crashes somewhere in wxWidgets 3.2.0:

```
Process:               GrandOrgue [677]
Path:                  /Users/USER/Documents/GrandOrgue.app/Contents/MacOS/GrandOrgue
Identifier:            com.our-organ.GrandOrgue
Version:               3.7.0 (3.7.0)
Code Type:             X86-64 (Native)
Parent Process:        zsh [649]
User ID:               502

Date/Time:             2022-08-09 09:29:49.283 -0700
OS Version:            Mac OS X 10.15.7 (19H15)
Report Version:        12
Anonymous UUID:        254FC794-BDEF-CB27-0AAC-9E1FFF82FA61


Time Awake Since Boot: 1200 seconds

System Integrity Protection: disabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000002, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Trace/BPT trap: 5
Termination Reason:    Namespace SIGNAL, Code 0x5
Terminating Process:   exc handler [677]

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libwx_baseu-3.2.0.0.0.dylib   	0x000000010248f293 wxDefaultAssertHandler(wxString const&, int, wxString const&, wxString const&, wxString const&) + 131
1   libwx_baseu-3.2.0.0.0.dylib   	0x000000010248d0f4 wxOnAssert(char const*, int, char const*, char const*, wchar_t const*) + 140
2   libwx_osx_cocoau_core-3.2.dylib	0x0000000101898fa0 wxArtProvider::GetBitmap(wxString const&, wxString const&, wxSize const&) + 402
3   libwx_osx_cocoau_core-3.2.dylib	0x0000000101899912 wxArtProvider::GetIcon(wxString const&, wxString const&, wxSize const&) + 26
4   libwx_osx_cocoau_core-3.2.dylib	0x00000001019a20d4 wxArtProvider::GetMessageBoxIcon(int) + 60
5   libwx_osx_cocoau_core-3.2.dylib	0x00000001019a5725 wxGenericMessageDialog::DoCreateMsgdialog() + 271
6   libwx_osx_cocoau_core-3.2.dylib	0x00000001019a5cfa wxGenericMessageDialog::ShowModal() + 48
7   libwx_osx_cocoau_core-3.2.dylib	0x0000000101897df2 wxGUIAppTraitsBase::ShowAssertDialog(wxString const&) + 480
8   libwx_baseu-3.2.dylib         	0x0000000101e0ea98 ShowAssertDialog(wxString const&, int, wxString const&, wxString const&, wxString const&, wxAppTraits*) + 689
9   libwx_baseu-3.2.dylib         	0x0000000101e0e6d1 wxAppConsoleBase::OnAssertFailure(wchar_t const*, int, wchar_t const*, wchar_t const*, wchar_t const*) + 205
10  libwx_baseu-3.2.0.0.0.dylib   	0x000000010248f318 wxDefaultAssertHandler(wxString const&, int, wxString const&, wxString const&, wxString const&) + 264
11  libwx_baseu-3.2.0.0.0.dylib   	0x000000010248d0f4 wxOnAssert(char const*, int, char const*, char const*, wchar_t const*) + 140
12  libwx_osx_cocoau_core-3.2.dylib	0x0000000101898fa0 wxArtProvider::GetBitmap(wxString const&, wxString const&, wxSize const&) + 402
13  com.our-organ.GrandOrgue      	0x00000001011f524d GOLogWindow::GOLogWindow(wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long) + 749
14  com.our-organ.GrandOrgue      	0x00000001011f55b9 GOLogWindow::GOLogWindow(wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long) + 25
15  com.our-organ.GrandOrgue      	0x00000001011f378e GOLog::GOLog(wxWindow*) + 286
16  com.our-organ.GrandOrgue      	0x00000001011f38ed GOLog::GOLog(wxWindow*) + 13
17  com.our-organ.GrandOrgue      	0x00000001010a593f GOApp::OnInit() + 1583
18  libwx_osx_cocoau_core-3.2.dylib	0x00000001017cac0c wxApp::CallOnInit() + 160
19  libwx_baseu-3.2.dylib         	0x0000000101e68200 wxEntry(int&, wchar_t**) + 65
20  com.our-organ.GrandOrgue      	0x00000001010a4726 main + 54
21  libdyld.dylib                 	0x00007fff715fbcc9 start + 1
```

**Comment by willeke1234** [comment]
With a lot of help from Google, I tried to get GitHub to build for macOS 10.15 on macOS 11. I'm not familiar with CMake, Homebrew and GitHub Actions. Results:

I modified the matrix in workflow `build.yml`:
`- run_on: macos-10.15`
to
`- run_on: macos-11`

and build script `build-on-osx.sh`:
`cmake -G "Unix Makefiles" $OS_PRMS $GO_PRMS . $SRC_DIR`
to
```
MACOSX_DEPLOYMENT_TARGET=10.15
cmake -G "Unix Makefiles" -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15  $OS_PRMS $GO_PRMS . $SRC_DIR
```
copied from [Can CMake specify the base SDK on Mac OS X?](https://stackoverflow.com/a/10559847)

GrandOrgue starts but immediately crashes with
`+[NSImage imageWithSystemSymbolName:accessibilityDescription:]: unrecognized selector sent to class 0x7fff8d9c8210`

	* thread #1, queue = 'com.apple.main-thread', stop reason = hit Objective-C exception
	  * frame #0: 0x00007fff6f711a6e libobjc.A.dylib`objc_exception_throw
		frame #1: 0x00007fff36c2f426 CoreFoundation`+[NSObject(NSObject) doesNotRecognizeSelector:] + 132
		frame #2: 0x00007fff36b14810 CoreFoundation`___forwarding___ + 1427
		frame #3: 0x00007fff36b141e8 CoreFoundation`__forwarding_prep_0___ + 120
		frame #4: 0x000000010175b2eb libwx_osx_cocoau_core-3.2.dylib`wxOSXGetSystemImage(wxString const&) + 48
		frame #5: 0x000000010175b348 libwx_osx_cocoau_core-3.2.dylib`wxOSXCreateSystemBitmapBundle(wxString const&, wxString const&, wxSize const&) + 17
		frame #6: 0x00000001017086ef libwx_osx_cocoau_core-3.2.dylib`wxMacArtProvider::CreateBitmapBundle(wxString const&, wxString const&, wxSize const&) + 23
		frame #7: 0x00000001017d4f51 libwx_osx_cocoau_core-3.2.dylib`wxArtProvider::GetBitmap(wxString const&, wxString const&, wxSize const&) + 323
		frame #8: 0x000000010015c8e3 GrandOrgue`GOLogWindow::GOLogWindow(wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long) + 787
		frame #9: 0x000000010015cc59 GrandOrgue`GOLogWindow::GOLogWindow(wxWindow*, int, wxString const&, wxPoint const&, wxSize const&, long) + 25
		frame #10: 0x000000010015af4d GrandOrgue`GOLog::GOLog(wxWindow*) + 253
		frame #11: 0x000000010015b0ad GrandOrgue`GOLog::GOLog(wxWindow*) + 13
		frame #12: 0x000000010000bccb GrandOrgue`GOApp::OnInit() + 1723
		frame #13: 0x0000000101706c0c libwx_osx_cocoau_core-3.2.dylib`wxApp::CallOnInit() + 160
		frame #14: 0x0000000101d9d200 libwx_baseu-3.2.dylib`wxEntry(int&, wchar_t**) + 65
		frame #15: 0x000000010000ab16 GrandOrgue`main + 54
		frame #16: 0x00007fff708b1cc9 libdyld.dylib`start + 1
		frame #17: 0x00007fff708b1cc9 libdyld.dylib`start + 1

`imageWithSystemSymbolName:accessibilityDescription:` is not available on 10.15. The wxWidgets source code:
```
WXImage wxOSXGetSystemImage(const wxString& name)
{
	wxCFStringRef cfname(name);

#if __MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_16
	if ( WX_IS_MACOS_AVAILABLE(11, 0) )
	{
		NSImage *symbol = [NSImage imageWithSystemSymbolName:cfname.AsNSString() accessibilityDescription:nil];
		if ( symbol )
			return symbol;
	}
#endif

	NSImage* nsimage = [NSImage imageNamed:cfname.AsNSString()];
	return nsimage;
}
```

`MAC_OS_X_VERSION_10_16` is 11. I don't see any checks in the debugger. Apparently wxWidgets is built for macOS 11 and the check is optimized away.
 
So I changed build script `prepare-osx.sh`:
`brew install gettext jack docbook-xsl wxmac cmake pkg-config fftw wavpack imagemagick`
to
`brew install gettext jack docbook-xsl wxwidgets@3.0 cmake pkg-config fftw wavpack imagemagick`
GrandOrgue runs but I didn't test everything.

I don't think `brew` can install wxWidgets (3.2) for macOS 10.15 on macOS 11. See [Homebrew libraries for deployment on older macOS versions #370](https://github.com/orgs/Homebrew/discussions/370)

**Reply by oleg68** [reply]
I don't think `brew` can install wxWidgets (3.2) for macOS 10.15 on macOS 11. See [Homebrew libraries for deployment on older macOS versions #370](https://github.com/orgs/Homebrew/discussions/370)

Most of Linux package managers can install packages from older releases. I think we have two options

1. Force brew to install wxWidgets 3.1 for 10.15 on OSx 11.
2. Download this package and install it manually. Is there a way of downloading brew packages?

**Reply by willeke1234** [reply]
`brew` can install wxWidgets 3.2  or 3.0. I tried `wxwidgets@3.1` but got an error:
>Warning: No available formula with the name "wxwidgets@3.1". Did you mean wxwidgets@3.0 or wxwidgets?

See https://formulae.brew.sh/formula/wxwidgets#default

I'm not familiar with brew, what I've understood from the linked discussion is
>homebrew seems to always install bottles for the current host operating system
>This is intentional.

No mention of a workaround.
See also https://github.com/Homebrew/discussions/discussions/2799

**Comment by oleg68** [comment]
@willeke1234 Using wxwidgets@3.0 is not desired: it does not support the dark theme.

**Reply by willeke1234** [reply]
`wxwidgets@3.0` was an experiment to check for other problems.

**Comment by oleg68** [comment]
I managed to build GrandOrgue on OSx 11 and to run it on OSx 10.15.

The steps are:
1. Using OSx 10.15, download the wxWidgets bottle for 10.15
    ```
     brew fetch wxwidgets
    ```
2. Save the downloaded bottle `90dab952e7e1c26dba572e1b19b7f97728cbbd4d5580a020aee0ff36eebe9138--wxwidgets--3.2.0.catalina.bottle.tar.gz` somewhere on a shared location
3. Using OSx 11, install all dependencies except wxwidgets
    ```
    brew install gettext jack docbook-xsl cmake pkg-config fftw wavpack imagemagick
    brew link gettext --force

    # A workaround of https://gitlab.kitware.com/cmake/cmake/-/issues/23826
    for F in $(grep -l '(, weak)\?'  /usr/local/Cellar/cmake/*/share/cmake/Modules/GetPrerequisites.cmake); do
      echo Patching $F...
      # OSx sed does not support the -i option
      sed -e 's/(, weak)\?/(, (weak|reexport))?/' $F >${F}.new
      mv ${F}.new $F
    done
    ```
5. Using OSx 11, install wxwidgets for 10.15
    ```
    brew install -f /Volumes/root/home/oleg/my-projects/grandorgue/wxmac/90dab952e7e1c26dba572e1b19b7f97728cbbd4d5580a020aee0ff36eebe9138--wxwidgets--3.2.0.catalina.bottle.tar.gz
    ```
6. Build GrandOrgue with the `-DCMAKE_OSX_DEPLOYMENT_TARGET=10.15` cmake parameter.

The built dmg image seems works on 10.15

**Comment by oleg68** [comment]
#1163 solves this problem. We continue to build GrandOrgue for 10.15.

**Comment by oleg68** [comment]
Now GO is built on MacOS 11. But for making GO compatible with Mac OS 10.15, I use the bottle wxwidgets--3.2.0.catalina.

Several days ago github builds stopped working: the attempt of installation of the bottle started failing:

```
% brew install -f --formula https://github.com/GrandOrgue/WxOsX/releases/download/3.2.0-1/90dab952e7e1c26dba572e1b19b7f97728cbbd4d5580a020aee0ff36eebe9138--wxwidgets--3.2.0.catalina.bottle.tar.gz
Error: can't modify frozen Hash
```

Googling this error didn't bring any useful information. Unfortunally, I'm not familiar with MacOS or homebrew, so I can'not fix it.

The easiest way woud be to install the current version of wxWidgets with brew instead the pinned one for 10.15. But it would break support of 10.15. 

Is it acceptable for users? If no, how to solve the build problems? Otherwise we are not able to build any releases on GitHub, neither for MacOs nor for other OS.

Especially I ask mac users, especially @willeke1234, for help.

**Reply by willeke1234** [reply]
I'm not really familiar with Homebrew. Supporting macOS 10.15 will be more and more difficult in the future. I would stop supporting macOS 10.15 now and put my time and effort into supporting Apple Silicon.

**Reply by MStraeten** [reply]
grandorgue builds fine on and for apple silicon (at least using macports) but needs setting the wavpack lib path manually 
See https://github.com/GrandOrgue/grandorgue/discussions/1439

**Comment by oleg68** [comment]
@willeke1234 thank you.

@rousseldenis @larspalo do you agree with stopping support of MacOs 10.15?

**Reply by larspalo** [reply]
@oleg68 Absolutely. Use whatever is currently available to get the builds up and running again!

**Reply by rousseldenis** [reply]
Yes, let's got forward
